### DISCUSSION TITLE
[Bug]: (some) strimzi PVC fails to resize

### DISCUSSION THREAD

**Discussion by leiyiz** [strimzi/strimzi-kafka-operator#8667] [General]
### Bug Description

the `CanSupport()` function in [csi_resizer.go](https://github.com/kubernetes-csi/external-resizer/blob/v1.4.0/pkg/resizer/csi_resizer.go#LL107C47-L107C47) has a line looking for 

```go
resizerName := pvc.Annotations[util.VolumeResizerKey]
```
and will fail the resize if that annotation pair does not exist and if the volume is a in-tree volume post migration. the `VolumeResizerKey` is defined [here](https://github.com/kubernetes-csi/external-resizer/blob/v1.4.0/pkg/util/events.go#L30) as `volume.kubernetes.io/storage-resizer` and I do not see this annotation in any of the PVCs created with strimzi. In a new cluster I created a in-tree gce-pd PVC and expanded it, after expansion I see the annotations:

```
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: pd.csi.storage.gke.io
               volume.kubernetes.io/storage-provisioner: pd.csi.storage.gke.io
               volume.kubernetes.io/storage-resizer: pd.csi.storage.gke.io
```

in the strimzi PVCs (no matter if the expansion succeeded or not) I see multiple annotations missing, including the storage-resizer annotation:
```
annotations:
    pv.kubernetes.io/bind-completed: "yes"
    strimzi.io/delete-claim: "false"
```

digging deeper I saw that after the annotation is added, `strimzi-cluster-operator` patched the PVC without these annotations and overwrote the PVC annotation which CSI migration depend on to correctly resize the PVC.

### Strimzi version

0.35.0

### Kubernetes version

1.23+

**Comment by scholzj** [comment]
I'm not sure I exactly follow the issue. The resizing works a little differently on every platform and Strimzi is platform agnostic. And I definitely don't think there is any general issue in resizing the volumes. Strimzi does not really know your environment and cannot set any custom annotations based on it. But if you want, you can add your own annotations through the template section of the Kafka resource: https://strimzi.io/docs/operators/latest/full/configuring.html#type-KafkaClusterTemplate-reference

**Reply by leiyiz** [reply]
I also don't know how resizing on other platforms work as I'm working on K8S specifically, but I guess my point is that strimzi shouldn't overwrite existing annotations set by kubernetes controllers. I see that strimzi controller is doing patch to k8s PVCs without including existing PVC annotations, by doing that those annotations are overwritten and in some situations k8s storage providers depend on them.

**Reply by scholzj** [reply]
> I also don't know how resizing on other platforms work as I'm working on K8S specifically

Platform in this case meant Google Cloud, not K8S.

> but I guess my point is that strimzi shouldn't overwrite existing annotations set by kubernetes controllers. I see that strimzi controller is doing patch to k8s PVCs without including existing PVC annotations, by doing that those annotations are overwritten and in some situations k8s storage providers depend on them.

There are many long discussions on that topic in other places. And the answer is not as easy as you suggest. 

But you can certainly add those annotations to the list in `PvcOperator` class to have them ignored. I do not think this is necessarily a problem. Please open a PR directly if you want or confirm if the list above is complete I will open a PR and get it into the next release.

Is this all that seems to be missing on GKE?
* `pv.kubernetes.io/bound-by-controller`
* `volume.beta.kubernetes.io/storage-provisioner`
* `volume.kubernetes.io/storage-provisioner`
* `volume.kubernetes.io/storage-resizer`

I think it should be no problem to add those.

**Reply by leiyiz** [reply]
Thanks for the response, by 
>  to have them ignored

do you mean that it's actually an intentional decision to overwrite the annotations on k8s objects? if that's the case, then yes let's get the annotation `volume.kubernetes.io/storage-resizer` ignored in that process.

`volume.kubernetes.io/storage-resizer` is the one causing issue with k8s csi migration (I think AWS also have the same problem as they are part of csi migration too so I didn't say just gke). However in general I think the other ones in the list are good to add too.

**Reply by scholzj** [reply]
I think I tested resizing on AWS with EBS CSI driver in 0.35 and it worked fine - but I do not remember the version of the driver and other details out of my head. Anyway, I will add all of them to the list. Thanks for raising this ... and sorry, I think I did not fully understand the issue after the first post. This indeed looks like a bug.

**Reply by leiyiz** [reply]
yeah, EBS CSI driver would be working fine, the issue here is that if you use in-tree driver in a stoageclass such as `aws-ebs` or `gce-pd` on a cluster version post-migration (say 1.26) then the migration code path kicks in (where I mentioned above the `CanSupport()` function - then the resize don't work because without annotation the CSI driver which is supposed to take over the in-tree pvc requests don't know it supports those PVCs

**Reply by leiyiz** [reply]
Thanks for all the discussion - I could've described the issue clearer

**Reply by scholzj** [reply]
Ok, I merged the fix. Thanks a lot for reporting this.

**Comment by msau42** [comment]
Thanks for fixing these specific annotations. In general though, I think it would be better to ignore all annotation keys with k8s.io or kubernetes.io suffixes. Kubernetes features may add new annotations so this would be better future proofing.

**Reply by scholzj** [reply]
There is a proposal in progress about moving to use server-side-apply, which should address this in general. That should also provide a better solution than just ignoring some annotations. 

Many of the kubernetes.io are driven from the Strimzi resources by the users and not set by some Kubernetes tooling. That does not seem to be the case here ... but for example the annotations for internal load balancers etc. So simply ignoring everything with the `kubernetes.io` domain in it would cause problems elsewhere.
